crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3706,16 +3707,44 @@ impl<'db> TypeInferenceBuilder<'db> {
             .member(db, &attr.id)
             .unwrap_with_diagnostic(|lookup_error| match lookup_error {
                 LookupError::Unbound => {
+                    let bound_on_instance = match value_type {
+                    Type::ClassLiteral(ClassLiteralType { class }) => {
+                        !class.instance_member(self.db(), attr).0.is_unbound()
+                    }
+                    Type::SubclassOf(subclass_of @ SubclassOfType { .. }) => {
+                        match subclass_of.subclass_of() {
+                            ClassBase::Class(class) => {
+                                !class.instance_member(self.db(), attr).0.is_unbound()
+                            }
+                            ClassBase::Dynamic(_) => unreachable!("For any symbol, lookup on a dynamic type will return a bound symbol of the same type"),
+                        }
+                    }
+                    _ => false,
+                };
+
+                if bound_on_instance {
+                    self.context.report_lint(
+                            &UNRESOLVED_ATTRIBUTE,
+                            attribute,
+                            format_args!(
+                                "Attribute `{}` can only be accessed on instances, not on the class object `{}` itself.",
+                                attr.id,
+                                value_type.display(self.db())
+                            ),
+                        );
```

Comment:
some overindentation here

```suggestion
                        &UNRESOLVED_ATTRIBUTE,
                        attribute,
                        format_args!(
                            "Attribute `{}` can only be accessed on instances, not on the class object `{}` itself.",
                            attr.id,
                            value_type.display(db)
                        ),
                    );
```